File: Java-Core/src/main/java/presto/grammar/MatchingExpressionConstraint.java
package presto.grammar;

import presto.error.InvalidDataError;
import presto.error.PrestoError;
import presto.expression.IExpression;
import presto.runtime.Context;
import presto.runtime.Variable;
import presto.type.AnyType;
import presto.utils.CodeWriter;
import presto.value.Boolean;
import presto.value.IValue;

public class MatchingExpressionConstraint implements IAttributeConstraint {

	IExpression expression;
	
	public MatchingExpressionConstraint(IExpression expression) {
		this.expression = expression;
	}
	
	@Override
	public void checkValue(Context context, IValue value) throws PrestoError {
		Context child = context.newChildContext();
		child.registerValue(new Variable("value", AnyType.instance()));
		child.setValue("value", value);
		Object test = expression.interpret(child);
		if(!Boolean.TRUE.equals(test))
			throw new InvalidDataError((value==null ? "null" : value.toString()) + " does not match:" + expression.toString());
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		writer.append(" matching ");
		expression.toDialect(writer);
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/grammar/MatchingExpressionConstraint.js
var InvalidDataError = require("../error/InvalidDataError").InvalidDataError;
var TransientVariable = require("../runtime/TransientVariable").TransientVariable;
var AnyType = require("../type/AnyType").AnyType;

function MatchingExpressionConstraint(expression) {
	this.expression = expression;
	return this;
}

MatchingExpressionConstraint.prototype.checkValue = function(context, value) {
	var child = context.newChildContext();
	child.registerValue(new TransientVariable("value", AnyType.instance));
	child.setValue("value", value);
	var test = this.expression.interpret(child);
	if(!test.value) {
		throw new InvalidDataError((value == null ? "null" : value.toString()) + " does not match:" + this.expression.toString());
	}
};

MatchingExpressionConstraint.prototype.toDialect = function(writer) {
    writer.append(" matching ");
    this.expression.toDialect(writer);
}

exports.MatchingExpressionConstraint = MatchingExpressionConstraint;

